src/library/scala/collection/TraversableLike.scala

```diff
@@ -246,11 +247,95 @@ trait TraversableLike[+A, +Repr] extends Any
   }
 
   private[scala] def filterImpl(p: A => Boolean, isFlipped: Boolean): Repr = {
-    val b = newBuilder
-    for (x <- this)
-      if (p(x) != isFlipped) b += x
+    this match {
+      case as: List[A] =>
+        filterImplList(as, p, isFlipped).asInstanceOf[Repr]
+      case _ =>
+        val b = newBuilder
+        for (x <- this)
+          if (p(x) != isFlipped) b += x
+
+        b.result
+    }
+  }
 
-    b.result
+  private[this] def filterImplList[A](self: List[A], p: A => Boolean, isFlipped: Boolean): List[A] = {
```

Comment:
should this be in the companion as it doesn't need a reference to `this`?

Reply:
I was seeking to do this in a way that wouldn't incur any grief from MiMa. Private methods in the companion object that are referred to from the class are JVM public.